lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
+    auto step = b.mapping.lookupOrDefault(forOp.getStep());
+    SmallVector<Value> initArgs;
+    for (auto idx : newLoopIndices) {
+      initArgs.push_back(forOp.getInitArgs()[idx]);
     }
-    // Store the result after the loop and load it after the base loop.
-    ImplicitLocOpBuilder b(result.getLoc(), baseLoop);
-    Value alloc = b.create<LocalAllocOp>(MemDescType::get(
-        ty.getShape(), ty.getElementType(), getSharedEncoding(ty),
-        SharedMemorySpaceAttr::get(ty.getContext()), /*mutable=*/true));
-    b.setInsertionPointAfter(loop);
-    b.create<LocalStoreOp>(result, alloc);
-    b.setInsertionPointAfter(baseLoop);
-    Value output = b.create<LocalLoadOp>(ty, alloc);
-    b.create<LocalDeallocOp>(alloc);
-    baseLoop.getResult(resultIdx).replaceAllUsesWith(output);
+    auto newForOp =
+        b.create<scf::ForOp>(forOp.getLoc(), lb, ub, step, initArgs);
+    newForOp->setAttrs(forOp->getAttrs());
+    newForOps.push_back(newForOp);
+
+    b.mapping.map(forOp.getInductionVar(), newForOp.getInductionVar());
+
+    auto oldIterArgs = forOp.getRegionIterArgs();
+    auto newIterArgs = newForOp.getRegionIterArgs();
+    for (auto [newIdx, oldIdx] : llvm::enumerate(newLoopIndices)) {
+      b.mapping.map(oldIterArgs[oldIdx], newIterArgs[newIdx]);
+      b.mapping.map(forOp.getResult(oldIdx), newForOp.getResult(newIdx));
+    }
+
+    b.setInsertionPointToStart(newForOp.getBody());
   }
 
-  // Delete everything else.
-  eraseOtherPartitions(loop, schedule, partition);
+  cloneOpsInBlock(forOp.getBody(), builders, schedule);
 
-  return std::move(block);
+  for (auto newForOp : newForOps) {
+    WarpSchedule::eraseFrom(newForOp);
+  }
 }
 
-//===----------------------------------------------------------------------===//
-// partitionLoop
-//===----------------------------------------------------------------------===//
+void mapRange(ValueRange fromRange, ValueRange toRange, IRMapping &mapping) {
+  for (auto [from, to] : llvm::zip(fromRange, toRange)) {
+    mapping.map(from, to);
+  }
+}
+
+SmallVector<size_t>
+getPartitionIndicesToCloneInto(const Partition *partition,
+                               const WarpSchedule &schedule) {
+  SmallVector<size_t> partitionIndices;
+
+  if (partition == schedule.getRootPartition()) {
+    for (size_t i = 0; i < schedule.getNumPartitions(); ++i) {
+      partitionIndices.push_back(i);
+    }
+  } else {
+    partitionIndices.push_back(partition->getIndex());
+  }
+
+  return partitionIndices;
+}
+
+void cloneIfOp(scf::IfOp ifOp, SmallVector<WarpGroupBuilder> &builders,
+               const WarpSchedule &schedule) {
+  auto partition = getPartition(ifOp, schedule);
+  assert(partition);
+  auto partitionIndices = getPartitionIndicesToCloneInto(partition, schedule);
+
+  SmallVector<scf::IfOp> newIfOps;
+  for (size_t idx : partitionIndices) {
+    auto &b = builders[idx];
+    auto cond = b.mapping.lookupOrDefault(ifOp.getCondition());
```

Comment:
nit: use `lookup` here instead

Reply:
Can't use that for the same reason as https://github.com/triton-lang/triton/pull/7415#discussion_r2205717616

if the condition is computed outside of the WS-ed loop, the lookup fails.